alacritty/src/display/damage.rs

```diff
@@ -1,20 +1,197 @@
+        if self.frames[0].damage_all {
+            vec![Rect::new(0, 0, size_info.width() as i32, size_info.height() as i32)]
+        } else {
+            let lines_damage = RenderDamageIterator::new(
+                TermDamageIterator::new(&self.frames[0].lines, 0),
+                &size_info,
+            );
+            lines_damage.chain(self.frames[0].rects.iter().copied()).collect()
+        }
+    }
+
+    /// Set the selection adding the damage for the current frame.
+    pub fn damage_selection(
+        &mut self,
+        mut selection: Option<SelectionRange>,
+        display_offset: usize,
+    ) {
+        mem::swap(&mut self.old_selection, &mut selection);
+
+        if self.frame().damage_all {
+            return;
+        }
+
+        if selection != self.old_selection {
+            for selection in self.old_selection.into_iter().chain(selection) {
+                // self.damage_selection_impl(selection, display_offset);
+                let display_offset = display_offset as i32;
+                let last_visible_line = self.screen_lines as i32 - 1;
+                let columns = self.columns;
+
+                // Don't damage invisible selection.
+                if selection.end.line.0 + display_offset < 0
+                    || selection.start.line.0.abs() < display_offset - last_visible_line
+                {
+                    continue;
+                };
+
+                let start = cmp::max(selection.start.line.0 + display_offset, 0);
+                let end = (selection.end.line.0 + display_offset).clamp(0, last_visible_line);
+                for line in start as usize..=end as usize {
+                    self.frame().lines[line].expand(0, columns - 1);
+                }
+            }
+        }
+    }
+}
+
+/// Damage state for the rendering frame.
+#[derive(Debug, Default)]
+pub struct FrameDamage {
+    /// The entire frame got redrawn.
+    damage_all: bool,
+    /// Terminal lines damaged in the given frame.
+    lines: Vec<LineDamageBounds>,
+    /// Rectangular regions damage in the given frame.
+    rects: Vec<Rect>,
+}
+
+impl FrameDamage {
+    /// Damage line for the given frame.
+    #[inline]
+    pub fn damage_line(&mut self, damage: LineDamageBounds) {
+        self.lines[damage.line].expand(damage.left, damage.right);
+    }
+
+    #[inline]
+    pub fn damage_point(&mut self, point: Point<usize>) {
+        self.lines[point.line].expand(point.column.0, point.column.0);
+    }
+
+    /// Mark the frame as fully damaged.
+    #[inline]
+    pub fn mark_fully_damaged(&mut self) {
+        self.damage_all = true;
+    }
+
+    /// Add a damage rectangle.
+    #[inline]
+    pub fn add_rect(&mut self, x: i32, y: i32, width: i32, height: i32) {
+        self.rects.push(Rect { x, y, width, height });
+    }
+
+    fn reset(&mut self, num_lines: usize, num_cols: usize) {
+        self.damage_all = false;
+        self.rects.clear();
+        self.lines.clear();
```

Comment:
I don't think this is necessary, but just to make sure: Do we have to deal with runaway memory here?

The line damage should always be limited to the viewport size which will be irrelevant for memory. But rects could potentially be a lot I'd imagine with a ton of underlines. I'd imagine the maximum number of rects you'd get in two frames wouldn't be enough to justify shrinking the vec on reset, but something to consider maybe.

Reply:
You're confusing which rects we're talking about. The damage rects here has nothing to do with underlines. It's for the UI elements outside of the viewport, so the amount is limited to 1-2 rectangles.

For now it's used in a single place with the message bar as you can see.